File: frontend/src/services/scenarioAccusation.ts
// =============================================================================
// scenarioAccusation.ts — client for the accusation section (task 2.11 B1)
// -----------------------------------------------------------------------------
// Endpoints:
//   GET    /api/cases/:slug/scenarios/:id/accusation
//   PUT    /api/cases/:slug/scenarios/:id/accusation
//   POST   /api/cases/:slug/scenarios/:id/accusation/instances/:nodeId
//   DELETE /api/cases/:slug/scenarios/:id/accusation/instances/:nodeId
//   PUT    /api/cases/:slug/scenarios/:id/accusation/instances/:nodeId/answer
//   DELETE /api/cases/:slug/scenarios/:id/accusation/instances/:nodeId/answer
//
// Same idioms as `scenarioAugmentation.ts`: `authFetch` (credentials + a 30s
// AbortController timeout), `encodeURIComponent` on every path segment,
// `readErrorMessage` to surface the backend's `{message}`, and every non-2xx
// throws (Standing Rule 1).
//
// ## The frontend composes nothing here
//
// `count_line` arrives as "Said 3 times, in 2 documents." and every gap arrives
// as a finished sentence with its `{code}` already in it. This file does not hold
// the templates they were built from, and could not rebuild them: that absence is
// the §10 exclusion made structural, not an omission. See
// backend/src/dto/scenario_accusation.rs.
//
// The one template that DOES arrive is `save_failed_template`, whose `{detail}`
// is the failure's own text — the single value that exists only on this side.
//
// The shapes mirror backend/src/dto/scenario_accusation.rs verbatim.
// =============================================================================

import { API_BASE_URL } from "./api";
import { authFetch } from "./auth";
import { readErrorMessage } from "./fetchUtils";

/** The stable gap tokens, mirroring `services::scenario_accusation_panel`. */
export const GAP_NO_ANSWER = "no_answer_prepared";
export const GAP_ACCUSATION_REMOVED = "accusation_removed";
export const GAP_ANSWER_REMOVED = "answer_removed";

export type AccusationInstanceDto = {
  graph_node_id: string;
  /** "C-14", or `null` for a candidate nothing has numbered yet. */
  code: string | null;
  answers_graph_node_id: string | null;
  answer_code: string | null;
  /** Whether the paired answer is still in the scenario (the Remove law). */
  answer_present: boolean;
};

export type AccusationGapDto = {
  /** One of the three `GAP_*` tokens. Branch on this, never on the message. */
  kind: string;
  graph_node_id: string;
  /** The stored sentence, already naming the fact it is about. */
  message: string;
};

export type AccusationWordingDto = {
  section_heading: string;
  section_hint: string;
  text_label: string;
  text_placeholder: string;
  text_missing_notice: string;
  text_edit_label: string;
  text_save_label: string;
  text_clear_label: string;
  text_cancel_label: string;
  mark_label: string;
  unmark_label: string;
  pair_label: string;
  repair_label: string;
  unpair_label: string;
  answer_label: string;
  picker_prompt: string;
  picker_cancel_label: string;
  /** Shown when the picker's filter leaves nothing. */
  picker_no_match_notice: string;
  /**
   * Shown when the picker had nothing to offer in the first place.
   *
   * Two sentences rather than one because the remedies differ — type something
   * else, versus there is nothing left to choose. Collapsing them would leave a
   * human retyping against an empty list (Standing Rule 1).
   */
  picker_empty_notice: string;
  gaps_heading: string;
  no_gaps_notice: string;
  /** Carries `{detail}` — filled by {@link fillDetail}. */
  save_failed_template: string;
};

export type AccusationPanelDto = {
  /** `null` until a human writes it; never stood in for. */
  accusation_text: string | null;
  /** "Said 3 times, in 2 documents." — `null` when nothing is marked. */
  count_line: string | null;
  /** The nothing-marked notice — `null` once anything is marked. */
  no_instances_notice: string | null;
  instances: AccusationInstanceDto[];
  gaps: AccusationGapDto[];
  wording: AccusationWordingDto;
};

function accusationUrl(slug: string, scenarioId: string): string {
  return (
    `${API_BASE_URL}/api/cases/${encodeURIComponent(slug)}` +
    `/scenarios/${encodeURIComponent(scenarioId)}/accusation`
  );
}

function instanceUrl(slug: string, scenarioId: string, graphNodeId: string): string {
  return `${accusationUrl(slug, scenarioId)}/instances/${encodeURIComponent(graphNodeId)}`;
}

/**
 * Put the failure's own words into the stored sentence.
 *
 * The one substitution this file performs, and it is not composition: the
 * template is the human's, and `{detail}` is the slot it leaves for something
 * only the browser knows. Mirrors `evidenceLinks.fillDetail`.
 */
export function fillDetail(template: string, detail: string): string {
  return template.replace("{detail}", detail);
}

/** Load the whole accusation section. */
export async function fetchAccusationPanel(
  slug: string,
  scenarioId: string,
): Promise<AccusationPanelDto> {
  const response = await authFetch(accusationUrl(slug, scenarioId));

  if (!response.ok) {
    const detail = await readErrorMessage(response);
    throw new Error(
      `Failed to load the accusation for scenario "${scenarioId}" ` +
        `(HTTP ${response.status}${detail}). Try again.`,
    );
  }

  const data: unknown = await response.json();
  const parsed = data as Partial<AccusationPanelDto>;

  // The shapes this section cannot render without. A missing `wording` would
  // leave every control unlabelled and a missing list would render as "nothing
  // marked" — indistinguishable from a scenario nobody has started, which is
  // exactly the confusion the honest-gap law exists to prevent.
  if (
    !Array.isArray(parsed.instances) ||
    !Array.isArray(parsed.gaps) ||
    parsed.wording == null
  ) {
    throw new Error(
      `The accusation payload for scenario "${scenarioId}" is missing ` +
        `instances/gaps/wording — backend/frontend contract mismatch. If this ` +
        `persists, report it to the site administrator.`,
    );
  }
  return parsed as AccusationPanelDto;
}

/**
 * Store the plain-words accusation, or withdraw it.
 *
 * `null` withdraws — a real act, which returns the rehearsal block to its named
 * gap. An empty string is NOT the way to withdraw: the backend refuses it,
 * because "withdraw this" and a slip of the keyboard are different intentions.
 */
export async function setAccusationText(
  slug: string,
  scenarioId: string,
  text: string | null,
): Promise<void> {
  const response = await authFetch(accusationUrl(slug, scenarioId), {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ text }),
  });

  if (!response.ok) {
    const detail = await readErrorMessage(response);
    throw new Error(
      `Failed to save the accusation for scenario "${scenarioId}" ` +
        `(HTTP ${response.status}${detail}).`,
    );
  }
}

/** Mark one included fact as an instance of the accusation. */
export async function markInstance(
  slug: string,
  scenarioId: string,
  graphNodeId: string,
): Promise<void> {
  const response = await authFetch(instanceUrl(slug, scenarioId, graphNodeId), {
    method: "POST",
  });

  if (!response.ok) {
    const detail = await readErrorMessage(response);
    throw new Error(
      `Failed to mark that statement as an accusation instance ` +
        `(HTTP ${response.status}${detail}).`,
    );
  }
}

/** Withdraw a marking. The fact stays in the scenario. */
export async function unmarkInstance(
  slug: string,
  scenarioId: string,
  graphNodeId: string,
): Promise<void> {
  const response = await authFetch(instanceUrl(slug, scenarioId, graphNodeId), {
    method: "DELETE",
  });

  if (!response.ok) {
    const detail = await readErrorMessage(response);
    throw new Error(
      `Failed to withdraw that accusation instance ` +
        `(HTTP ${response.status}${detail}).`,
    );
  }
}

/**
 * Pair — or re-pair — the record item that answers one instance.
 *
 * Re-pairing REPLACES: one answer per accusation, enforced by a partial unique
 * index. Sending it twice leaves the same one stored, which is why this is a PUT.
 */
export async function pairAnswer(
  slug: string,
  scenarioId: string,
  graphNodeId: string,
  answersGraphNodeId: string,
): Promise<void> {
  const response = await authFetch(`${instanceUrl(slug, scenarioId, graphNodeId)}/answer`, {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ answers_graph_node_id: answersGraphNodeId }),
  });

  if (!response.ok) {
    const detail = await readErrorMessage(response);
    throw new Error(
      `Failed to pair that answer to the accusation instance ` +
        `(HTTP ${response.status}${detail}).`,
    );
  }
}

/** Remove the answer paired to one instance, returning it to the prep list. */
export async function unpairAnswer(
  slug: string,
  scenarioId: string,
  graphNodeId: string,
): Promise<void> {
  const response = await authFetch(`${instanceUrl(slug, scenarioId, graphNodeId)}/answer`, {
    method: "DELETE",
  });

  if (!response.ok) {
    const detail = await readErrorMessage(response);
    throw new Error(
      `Failed to un-pair that answer (HTTP ${response.status}${detail}).`,
    );
  }
}

File: backend/src/dto/scenario_accusation.rs
//! Wire shapes for authoring the accusation and its answers (task 2.11 B1).
//!
//! One read payload and two request bodies. The read is what the working view's
//! accusation section renders; the requests are what its three write concerns
//! send.
//!
//! ## Every sentence arrives composed, and the templates deliberately do not
//!
//! The count line, the three gap messages and the no-instances notice are built
//! on THIS side of the wire and cross it as finished sentences. The templates
//! they were built from are not sent at all.
//!
//! That is not tidiness, it is the §10 exclusion made structural. A browser
//! holding `accusation_count_template` would need the instance list and the
//! document map to fill it, and would then be holding exactly the data the
//! rehearsal page is forbidden to render. A browser that receives "Said 3 times,
//! in 2 documents." can render that and nothing else.
//!
//! The one template that DOES cross is [`AccusationWordingDto::save_failed_template`],
//! because its `{detail}` is the failure's own text — the single value that exists
//! only in the browser.
//!
//! ## Domain note: no verdicts, no confidence, no notes, no internal ids
//!
//! Nothing here carries a tier, an ordinal, a confidence, a status, a human's
//! annotation, or a database id. The anchors are graph node ids (§12) because that
//! is what a judgment is keyed to and what a click sends back; everything else a
//! human reads about a fact they already have from the cards payload.

use serde::{Deserialize, Serialize};

/// One marked instance, and whatever a human paired to it.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct AccusationInstanceDto {
    /// The statement that IS them making the accusation.
    pub graph_node_id: String,
    /// The handle a human calls it by — `"C-14"`, or `null` for a candidate
    /// nothing has numbered yet. `null` rather than the node id: a code that is
    /// really an id would read as a handle and be quoted as one out loud.
    pub code: Option<String>,
    /// The record item paired as our answer, or `null` when nobody has paired one.
    pub answers_graph_node_id: Option<String>,
    /// That answer's handle, on the same terms as `code`.
    pub answer_code: Option<String>,
    /// Whether the paired answer is still in the scenario.
    ///
    /// `false` with a non-null `answers_graph_node_id` is the Remove law's visible
    /// broken pairing — the row is kept, and the matching gap below says so. The
    /// client is TOLD this rather than asked to infer it by hunting the gap list,
    /// for the same reason `status` travels beside `status_label`.
    pub answer_present: bool,
}

/// One named absence, ready to render.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct AccusationGapDto {
    /// A stable token — `no_answer_prepared`, `accusation_removed`,
    /// `answer_removed`.
    ///
    /// ## Why a token travels beside the sentence
    ///
    /// So the client can treat the prep list differently from a broken pairing
    /// (louder, first) WITHOUT parsing prose. A client that decided by matching on
    /// "NO ANSWER" would silently stop distinguishing them the first time Roman
    /// edited that row — which he is invited to do.
    pub kind: String,
    /// The instance the gap is about.
    pub graph_node_id: String,
    /// The stored sentence with its `{code}` already filled in.
    pub message: String,
}

/// Every word the accusation section renders that is not already a sentence.
///
/// A wire mirror of `domain::wording_accusation::AccusationWording` MINUS the six
/// templates the server fills — see this module's header for why their absence is
/// the point.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct AccusationWordingDto {
    pub section_heading: String,
    pub section_hint: String,
    pub text_label: String,
    pub text_placeholder: String,
    pub text_missing_notice: String,
    pub text_edit_label: String,
    pub text_save_label: String,
    pub text_clear_label: String,
    pub text_cancel_label: String,
    pub mark_label: String,
    pub unmark_label: String,
    pub pair_label: String,
    pub repair_label: String,
    pub unpair_label: String,
    pub answer_label: String,
    pub picker_prompt: String,
    pub picker_cancel_label: String,
    /// Shown when the picker's filter leaves nothing.
    pub picker_no_match_notice: String,
    /// Shown when the picker had nothing to offer in the first place. A different
    /// state with a different remedy, so never the same sentence.
    pub picker_empty_notice: String,
    pub gaps_heading: String,
    pub no_gaps_notice: String,
    /// Carries `{detail}` — the one placeholder only the browser can fill.
    pub save_failed_template: String,
}

/// The accusation section in one read.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct AccusationPanelDto {
    /// The standing accusation in a human's plain words, or `null`.
    ///
    /// `null` is the honest-gap law: nobody has written it, so the section shows
    /// `wording.text_missing_notice` and never a quote from the record standing in
    /// for it. It is never derived from `attack_text`.
    pub accusation_text: Option<String>,
    /// "Said 3 times, in 2 documents." — `null` when nothing is marked, because a
    /// count of zero is not a count, it is the notice below.
    pub count_line: Option<String>,
    /// "No instances marked yet. 46 included facts are waiting here." — `null`
    /// once anything is marked. Exactly one of this and `count_line` is present,
    /// which is what stops the section rendering both or neither.
    pub no_instances_notice: Option<String>,
    /// The marked instances, in the derivation's total order.
    pub instances: Vec<AccusationInstanceDto>,
    /// Every named absence, prep list first.
    pub gaps: Vec<AccusationGapDto>,
    pub wording: AccusationWordingDto,
}

/// Body of the accusation-sentence write.
///
/// ## Why `Option<String>` and not two routes
///
/// `null` CLEARS the sentence, and clearing is a real act — a human withdrawing
/// words they no longer stand behind, which returns the rehearsal block to its
/// named gap. Modelling it as an absent field on the same route keeps one place
/// where the sentence is decided; a separate DELETE would be a second write path
/// to fence.
///
/// An empty or whitespace-only string is REFUSED rather than treated as a clear.
/// They are different intentions — one is "withdraw this", the other is a slip —
/// and the column's CHECK refuses a blank anyway, so accepting it here would only
/// change which layer said no.
#[derive(Debug, Clone, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct SetAccusationTextRequest {
    /// The sentence, or `null` to withdraw it.
    ///
    /// ## Rust Learning: `#[serde(default)]` on an `Option` field
    ///
    /// Without it, `{}` is a deserialization ERROR ("missing field text") while
    /// `{"text": null}` succeeds. With it, both mean `None`. Both spellings mean
    /// "no sentence" to a human, so both are accepted — and `deny_unknown_fields`
    /// still catches the typo (`{"txet": "…"}`) that would otherwise clear the
    /// accusation silently.
    #[serde(default)]
    pub text: Option<String>,
}

/// Body of the answer pairing.
///
/// One field, required: a pairing with no answer in it is not a pairing, and the
/// database says so too (`scenario_human_facts_answer_needs_anchor`). Unpairing
/// has its own DELETE rather than being a `null` here, because "this no longer
/// answers it" and "this answers it instead" are different acts and the log has to
/// tell them apart.
#[derive(Debug, Clone, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct PairAnswerRequest {
    /// The record item that answers the anchored instance.
    pub answers_graph_node_id: String,
}
